// Copyright 2011, Tencent Inc.
// Author: Hsiaokang Liu <hsiaokangliu@tencent.com>
//   Dongping HUANG <dphuang@tencent.com>
//   Eric Liu <ericliu@tencent.com>

package poppy;

import "poppy/rpc_option.proto";
import "poppy/rpc_message.proto";

option java_package="com.soso.poppy";

message HealthResponse {
    required string health = 1;
}

message MethodDescriptor {
    required string name = 1;
}

message ServiceDescriptor {
    required string name = 1;
    repeated MethodDescriptor methods = 2;
}

message EnumValueDescriptor {
    required string name = 1;
    required int32 number = 2;
}

message EnumDescriptor {
    required string name = 1;
    repeated EnumValueDescriptor values = 2;
}

message FieldDescriptor {
    required string name = 1;
    required int32 number = 2;
    required int32 label = 3;
    required int32 type = 4;
    optional string type_name = 5;
    optional string default_value = 6;
}

message MessageDescriptor {
    required string name = 1;
    repeated FieldDescriptor fields = 2;
}

message AllServiceDescriptors {
    repeated ServiceDescriptor services = 1;
}

message MethodInputTypeDescriptorsRequest {
    required string method_full_name = 1;
}

message MethodInputTypeDescriptorsResponse {
    required string request_type = 1;
    repeated MessageDescriptor message_types = 2;
    repeated EnumDescriptor enum_types = 3;
}

// The server side serivce.
service BuiltinService {
    // The service timeout is 3 seconds.
    option (poppy.service_timeout) = 3000;

    rpc Health(EmptyRequest) returns(HealthResponse);
    rpc GetAllServiceDescriptors(EmptyRequest)
            returns(AllServiceDescriptors);
    rpc GetMethodInputTypeDescriptors(MethodInputTypeDescriptorsRequest)
            returns(MethodInputTypeDescriptorsResponse);
}
